### PR TITLE
token_to_id being list, type annotations

### PR DESCRIPTION
**Pull request by allenoxli**

### CODE REVIEW

File: example_generate.py [hunk 1]
```diff
@@ -2,27 +2,39 @@
...
 import lmp
 
+
 #####################################################################
 # 檔案路徑設定
 #####################################################################
-experiment_no = 1
+
+# experiment_no = 1
+experiment_no = 5
```
Comment (ProFatXuanAll) [inline example_generate.py]: remove unnecessary comment and change it to `experiment_no = 1`

File: example_generate.py [hunk 2]
```diff
@@ -2,27 +2,39 @@
...
 config = lmp.config.BaseConfig.load_from_file(f'{model_path}/config.pickle')
-
 tokenizer = lmp.tokenizer.CharTokenizer.load_from_file(f'{model_path}/tokenizer.pickle')
 
-model = lmp.model.GRUModel(config=config,
+# tokenizer = lmp.tokenizer.CharTokenizerDict.load_from_file(f'{model_path}/tokenizer.pickle')
+
+# model = lmp.model.GRUModel(config=config,
+#                            tokenizer=tokenizer)
+model = lmp.model.LSTMModel(config=config,
```
Comment (ProFatXuanAll) [inline example_generate.py]: remove unnecessary comment

File: example_generate.py [hunk 3]
```diff
@@ -2,27 +2,39 @@
...
                            tokenizer=tokenizer)
 model.load_state_dict(torch.load(f'{model_path}/model.ckpt'))
 
+
+
+
+BOS = sys.argv[1] if len(sys.argv) > 1 and sys.argv[1] != ''  else '今天'
```
Comment (ProFatXuanAll) [inline example_generate.py]: remove unnecessary code

File: example_train.py [hunk 4]
```diff
@@ -11,14 +11,15 @@
...
+import sys
 
 # self-made modules
 import lmp
 
 ##############################################
 # Hyperparameters setup
 ##############################################
-experiment_no = 1
+experiment_no = 6
```
Comment (ProFatXuanAll) [inline example_train.py]: change it to `experiment_no = 1`

File: example_train.py [hunk 5]
```diff
@@ -57,9 +58,13 @@
 ##############################################
 tokenizer = lmp.tokenizer.CharTokenizer()
 
+# 讓使用者決定是否 uncase
+uncaseOrNot = True if len(sys.argv) > 1 and sys.argv[1].lower() == 'uncase'   else False
```
Comment (ProFatXuanAll) [inline example_train.py]: rewrite the script and use built-in `argparse`

File: example_train.py [hunk 6]
```diff
@@ -70,8 +75,11 @@
 ##############################################
 # Construct RNN model, choose loss function and optimizer.
 ##############################################
-model = lmp.model.GRUModel(config=config,
+model = lmp.model.LSTMModel(config=config,
                            tokenizer=tokenizer)
+# model = lmp.model.GRUModel(config=config,
```
Comment (ProFatXuanAll) [inline example_train.py]: remove unnecessary comment

File: lmp/config.py [hunk 7]
```diff
@@ -3,21 +3,31 @@
...
-        self.learning_rate = kwargs.pop('learning_rate', 10e-4)
-        self.min_count = kwargs.pop('min_count', 0)
-        self.num_rnn_layers = kwargs.pop('num_rnn_layers', 1)
-        self.num_linear_layers = kwargs.pop('num_linear_layers', 1)
-        self.seed = kwargs.pop('seed', 1)
+    def __init__(self, batch_size:int=1,
+                     dropout:float=0,
+                     embedding_dim:int=1,
+                     epoch:int=1,
+                     grad_clip_value:float=1,
```
Comment (ProFatXuanAll) [inline lmp/config.py]: change `grad_clip_value` to `max_norm`.

File: lmp/config.py [hunk 8]
```diff
@@ -41,7 +51,7 @@ def load_from_file(cls, file_path=None):
 
         return self
 
-    def save_to_file(self, file_path=None):
+    def save_to_file(self, file_path:str=None):
```
Comment (ProFatXuanAll) [inline lmp/config.py]: add space after colon

File: lmp/config.py [hunk 9]
```diff
@@ -3,21 +3,31 @@
...
-        self.embedding_dim = kwargs.pop('embedding_dim', 1)
-        self.epoch = kwargs.pop('epoch', 1)
-        self.grad_clip_value = kwargs.pop('grad_clip_value', 1)
-        self.hidden_dim = kwargs.pop('hidden_dim', 1)
-        self.learning_rate = kwargs.pop('learning_rate', 10e-4)
-        self.min_count = kwargs.pop('min_count', 0)
-        self.num_rnn_layers = kwargs.pop('num_rnn_layers', 1)
-        self.num_linear_layers = kwargs.pop('num_linear_layers', 1)
-        self.seed = kwargs.pop('seed', 1)
+    def __init__(self, batch_size:int=1,
```
Comment (ProFatXuanAll) [inline lmp/config.py]: change style to
```py
def __init__(
        self,
        batch_size: int = 1,
        ...
):
```
add space after colon

File: lmp/dataset.py [hunk 10]
```diff
@@ -1,29 +1,39 @@
 import torch.utils.data
 import torch.nn.utils.rnn
+from typing import List, Tuple, Union
+import lmp.tokenizer
+import lmp.config
+
 
 class BaseDataset(torch.utils.data.Dataset):
-    def __init__(self, **kwargs):
+    def __init__(self, text_list: List[str],
```
Comment (ProFatXuanAll) [inline lmp/dataset.py]: change style to
```py
def __init__(
        self,
        text_list: List[str],
        ...
):
```
add space after colon

File: lmp/dataset.py [hunk 11]
```diff
@@ -1,29 +1,39 @@
...
 
-        self.text_list = kwargs.pop('text_list', [])
-        self.config = kwargs.pop('config', None)
-        self.tokenizer = kwargs.pop('tokenizer', None)
-
-        self.tokenizer.build_dict(self.text_list, self.config.min_count)
+        self.text_list = text_list
+        self.config = config
+        self.tokenizer = tokenizer
+        self.uncase = uncase  # 是否把大小寫視為不同字, default: False
```
Comment (ProFatXuanAll) [inline lmp/dataset.py]: when value of variable is `bool` type, named the variable using prefix `is_`.
for example: `is_uncased`

File: lmp/tokenizer.py [hunk 12]
```diff
@@ -2,24 +2,193 @@
...
 import re
+import abc
+from typing import List, Dict
+
+# token_to_id 的 type 為 list(省記憶體)
+
 
 class BaseTokenizer:
-    def __init__(self, **kwargs):
+    def __init__(self, pad_token: str = '[PAD]', pad_token_id: int = 0,
```
Comment (ProFatXuanAll) [inline lmp/tokenizer.py]: Name the classes `BaseTokenizerByList` and `BaseTokenizerByDict`

File: lmp/tokenizer.py [hunk 13]
```diff
@@ -2,24 +2,193 @@
...
+        self.eos_token = eos_token
+        self.eos_token_id = eos_token_id
         # unknown token
-        self.unk_token = kwargs.pop('unk_token', '[UNK]')
-        self.unk_token_id = kwargs.pop('unk_token_id', 4)
+        self.unk_token = unk_token
+        self.unk_token_id = unk_token_id
+
+        self.token_to_id: List[str] = []  # 用 list 節省記憶體
+        self.id_to_token: Dict[int, str] = {}
```
Comment (ProFatXuanAll) [inline lmp/tokenizer.py]: inverse lookup should only use `self.token_to_id.index`

File: lmp/tokenizer.py [hunk 14]
```diff
@@ -2,24 +2,193 @@
...
+    def convert_tokens_to_sentences(self, all_tokens: List[List[str]]) -> List[str]:
+        return [self.detokenize(tokens) for tokens in all_tokens]
+
+    def convert_tokens_to_ids(self, all_tokens: List[List[str]]) -> List[List[int]]:
+        result = []
+        for tokens in all_tokens:
+            ids = []
+            for token in tokens:
+                if token in self.token_to_id:
+                    # 取出該token在 list的index
```
Comment (ProFatXuanAll) [inline lmp/tokenizer.py]: prepend and append space around english words

File: example_generate.py [hunk 15]
```diff
@@ -11,29 +10,23 @@
 # 檔案路徑設定
 #####################################################################
 
-# experiment_no = 1
-experiment_no = 5
+experiment_no = 1
```
Comment (ProFatXuanAll) [inline example_generate.py]: use `argparse` to get `experiment_no`, wrap whole generation script into a function.

File: example_train.py [hunk 16]
```diff
@@ -59,12 +59,15 @@
 tokenizer = lmp.tokenizer.CharTokenizer()
 
 # 讓使用者決定是否 uncase
-uncaseOrNot = True if len(sys.argv) > 1 and sys.argv[1].lower() == 'uncase'   else False
+parser = argparse.ArgumentParser()
+parser.add_argument("-u", "--uncase", help="regard capital letter and lowercase letter as same word",
```
Comment (ProFatXuanAll) [inline example_train.py]: Do not use short-hand argument (i.e. `"-u"`).
Change helper text into `"convert all upper case into lower case."`.
Indicate `--uncase` as required argument.

File: example_train.py [hunk 17]
```diff
@@ -59,12 +59,15 @@
...
+                    action="store_true")
+args = parser.parse_args()
 
 dataset = lmp.dataset.BaseDataset(config=config,
                                   text_list=df['title'],
                                   tokenizer=tokenizer,
-                                  uncase=uncaseOrNot)
+                                  is_uncased=args.uncase)
```
Comment (ProFatXuanAll) [inline example_train.py]: Put `uncase` into `config` object.
Parse all config option by `argparse`.

File: example_train.py (hunk 16 above)
Comment (ProFatXuanAll) [reply example_train.py]: rename `arg.uncase` into `arg.is_uncased` for consistency

File: lmp/tokenizer.py [hunk 18]
```diff
@@ -135,8 +130,8 @@ def decode(self, all_ids: List[List[int]]) -> List[str]:
 
         return result
 
-    def build_dict(self, all_sentences: List[str], min_count: int = 0, uncase: bool = False):
-        if uncase:  # 如果uncase是True，就不分大小寫
+    def build_dict(self, all_sentences: List[str], min_count: int = 0, is_uncased: bool = False):
+        if is_uncased:  # 如果 is_uncased 是 True，就不分大小寫
```
Comment (ProFatXuanAll) [inline lmp/tokenizer.py]: Write docstring

Comment (allenoxli) [reply lmp/tokenizer.py]: 我需要把每個 function和class 都寫 docstring ?

File: example_train.py [hunk 19]
```diff
@@ -11,26 +19,57 @@
...
+def boolean_string(s):
+    if s not in {'False', 'True'}:
+        raise ValueError('Not a valid boolean string')
+    return s == 'True'
+
+parser = argparse.ArgumentParser()
+
+# Required arguments.
+parser.add_argument("--experiment_no", type=int, default=1, required=True, help="using which experiment_no data")
+parser.add_argument("--is_uncased",    type=boolean_string, default=False, required=True, help="convert all upper case into lower case.")
```
Comment (ProFatXuanAll) [inline example_train.py]: Using `--is_uncased` without passing value indicate `True`.
See https://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse

File: example_train.py [hunk 20]
```diff
@@ -11,26 +19,57 @@
...
+
+# Optional arguments.
+parser.add_argument("--batch_size",         type=int,   default=32,     help="Training batch size.")
+parser.add_argument("--dropout",            type=float, default=0,      help="Dropout rate.")
+parser.add_argument("--embedding_dim",      type=int,   default=100,    help="Embedding dimension.")
+parser.add_argument("--epoch",              type=int,   default=10,     help="Number of training epochs.")
+parser.add_argument("--max_norm",           type=float, default=1,      help="Max norm of gradient.Used when cliping gradient norm.")
+parser.add_argument("--hidden_dim",         type=int,   default=300,    help="Hidden dimension.")
+parser.add_argument("--learning_rate",      type=float, default=10e-4,  help="Optimizer's parameter `lr`.")
```
Comment (ProFatXuanAll) [inline example_train.py]: `10e-4 == 1e-3`, use the latter one.
This must be consider as bugs.

Comment (ProFatXuanAll) [reply example_train.py]: Also change the default value to `5e-5`

File: example_train.py [hunk 21]
```diff
@@ -11,26 +19,57 @@
...
+import argparse
 
 # self-made modules
 import lmp
 
+def boolean_string(s):
+    if s not in {'False', 'True'}:
+        raise ValueError('Not a valid boolean string')
+    return s == 'True'
+
```
Comment (ProFatXuanAll) [inline example_train.py]: wrap all `argparser` code into scope `if __name__ == '__main__':`

File: example_train.py [hunk 22]
```diff
@@ -11,26 +19,57 @@
...
-                               epoch=10,
-                               grad_clip_value=1,
-                               hidden_dim=300,
-                               learning_rate=10e-4,
-                               min_count=0,
-                               num_rnn_layers=1,
-                               num_linear_layers=2,
-                               seed=7)
+experiment_no = args.experiment_no
+config = lmp.config.BaseConfig(batch_size=args.batch_size,
```
Comment (ProFatXuanAll) [inline example_train.py]: Pass argument in alphabetic order.
Style code into
```py
lmp.config.BaseConfig(
    batch_size=...,
    ...
)
```

File: example_train.py [hunk 23]
```diff
@@ -56,15 +95,13 @@
 ##############################################
 # Construct tokenizer and perform tokenization.
 ##############################################
-tokenizer = lmp.tokenizer.CharTokenizer()
+tokenizer = lmp.tokenizer.CharTokenizerByList()
```
Comment (ProFatXuanAll) [inline example_train.py]: Tokenizer should be construct through `argparser`

File: example_train.py [hunk 24]
```diff
@@ -101,24 +136,36 @@
...
+
 best_loss = None
+
+model_num = 0
 for epoch in range(config.epoch):
-    model.train()
+    
     print(f'epoch {epoch}')
     total_loss = 0
+    iteration_times = 0
```
Comment (ProFatXuanAll) [inline example_train.py]: Change `iteration_times` into `step`

File: example_train.py [hunk 25]
```diff
@@ -77,8 +114,6 @@
 ##############################################
 model = lmp.model.LSTMModel(config=config,
```
Comment (ProFatXuanAll) [inline example_train.py]: `model` should be construct through `argparser`

File: example_generate.py [hunk 26]
```diff
@@ -11,30 +18,37 @@
...
-model_path = f'{data_path}/{experiment_no}'
 
-config = lmp.config.BaseConfig.load_from_file(f'{model_path}/config.pickle')
-tokenizer = lmp.tokenizer.CharTokenizer.load_from_file(f'{model_path}/tokenizer.pickle')
+def generate_sentences(experiment_no=1):
+    data_path = os.path.abspath('./data')
+    model_path = f'{data_path}/{experiment_no}'
+
+    config = lmp.config.BaseConfig.load_from_file(f'{model_path}/config.pickle')
+    tokenizer = lmp.tokenizer.CharTokenizerByList.load_from_file(f'{model_path}/tokenizer.pickle')
```
Comment (ProFatXuanAll) [inline example_generate.py]: Use `config` to construct `tokenizer` and `model`.

File: test.py [hunk 27]
```diff
@@ -0,0 +1,10 @@
+import argparse
```
Comment (ProFatXuanAll) [inline test.py]: Remove unused code.
If this file is intended to perform unit test, then use `unittest` or other framework to write unit test

**Review by ProFatXuanAll — APPROVED** [review]

File: lmp/config.py [hunk 28]
```diff
@@ -108,7 +108,7 @@ def __init__(
             num_rnn_layers: int = 1,
             optimizer_class: str = 'adam',
             seed: int = 1,
-            tokenizer_class: str = 'list'
+            tokenizer_class: str = 'char_dict''
```
Comment (ProFatXuanAll) [inline lmp/config.py]: syntax error

**Review by ProFatXuanAll — APPROVED** [review]

**Pull request merged**